influxdata/influxdb
# Delete series support for TSM

###### Required for all non-trivial PRs
- [x] Rebased/mergable

This PR adds the capability to the TSM engine to remove points based on time ranges to support the `DELETE SERIES` statement.  The query parser/execution is not part of this PR.

The main changes to support this capability are:

* Changed the tombstone file format to also store time ranges per key
* Update `indirectIndex` to support partially deleted keys
* Update WAL to store time ranges with deleted keys
* Update Cache to allow removing keys by ranges of time
* Update CacheLoader to handle keys that are partially deleted
* Update Compactor to skip and filter out blocks and values that are partially deleted

There are a few unrelated changes in the PR done to make the implementation simpler as well as fix some bugs:

* `directIndex` and `indirectIndex` implemented the same `TSMIndex` interface which is used for reading and writing.  We really only used `directIndex` for writing and tests and `indirectIndex` for reading.  These have been split out into separate reader and writer types so that we do not need to continue to implement reading capabilities in the `directIndex` when they are not used.  This will also make it easier to support different version of indexes in the future when we need to extend the TSM file format.
* The prior change allowed for a lot of code to be removed (mainly `fileAccessor`)
* Fixed many places where corrupted WAL segments could panic.  Should fix #6468 and similar ones in the future.
* Removed `NewTSMReaderWithOptions` which allowed creating a TSMReader w/ a direct or indirect index.  There is only `NewTSMReader` now that always uses `indirectIndex`.  Create a `TSMReader` with a `directIndex` was only useful for writing tests that don't write to disk.
* Removed several inefficient `TSMReader` funcs that should never be called (e.g. `Keys`)

cc @benbjohnson
